src/Encoder.h

```diff
@@ -10,33 +17,59 @@ namespace cppkin
 {
     class Span;
     class SimpleAnnotation;
-    class EncoderContext;
 
-    template<EncodingTypes::Enumeration >
-    class Encoder
+    class Encoder {
+    public:
+        virtual ~Encoder() {}
+        virtual std::string ToString(const Span&) const = 0;
+        virtual std::string ToString(const std::vector<EncoderContext::ContextElement>&) const = 0;
+        virtual std::string ToString(const std::vector<Span*>&) const = 0;
+    };
+
+    template<EncodingTypes::Enumeration>
+    class EncoderImpl: public Encoder
     {
     public:
-        static void Serialize(EncoderContext& context, const Span::SpanHeader& spanHeader){
+        EncoderImpl();
+        virtual std::string ToString(const Span&) const {
             throw core::Exception(SOURCE, "Unsupported method");
         }
-        static void Serialize(EncoderContext& context, const Span& span){
+
+        virtual std::string ToString(const std::vector<EncoderContext::ContextElement>&) const {
             throw core::Exception(SOURCE, "Unsupported method");
         }
-        static Span::SpanHeader DeSerializeSpanHeader(EncoderContext& context){
+
+        virtual std::string ToString(const std::vector<Span*>&) {
             throw core::Exception(SOURCE, "Unsupported method");
         }
     };
 
     template<>
-    class Encoder<EncodingTypes::Thrift>
+    class EncoderImpl<EncodingTypes::Thrift>: public Encoder
     {
     public:
-        static void Serialize(EncoderContext& context, const Span::SpanHeader& spanHeader);
-        static void Serialize(EncoderContext& context, const Span& span);
-        static Span::SpanHeader DeSerializeSpanHeader(EncoderContext& context);
+        EncoderImpl();
+        virtual std::string ToString(const Span&) const;
+        virtual std::string ToString(const std::vector<EncoderContext::ContextElement>&) const;
+        virtual std::string ToString(const std::vector<Span*>&) const;
 
     private:
-        static void Serialize(::Span& thriftSpan, const SimpleAnnotation& annotation);
+        template <class T>
+        std::string SpansVectorToString(T& spans) const;
```

Comment:
Should be the impel of ToString(const std::vector<ContextElement>&) if we will agree on removing the ToString(const std::vector<Span*>&) method.